### PR TITLE
webkit2gtk: update to 2.38.1.

### PR DESCRIPTION
**Pull request by oreo639**
[ci skip]

This splits webkit2gtk into 3 packages:
+ webkit2gtk40: GTK+3 and libsoup2
+ webkit2gtk41: GTK+3 and libsoup3
+ webkit2gtk50: GTK4 and libsoup3

This is necessary since libsoup2 and libsoup3 cannot be used by an application at the same time.
Ideally in the future, the packages using webkit2gtk40 would be moved to webkit2gtk41 or 50 (or be dropped).

I tested this PR with epiphany 42 and it seems to work (and it compiles sucessfully).

#### Testing the changes
- I tested the changes in this PR: **briefly**

### CODE REVIEW

File: srcpkgs/webkit2gtk/template [hunk 1]
```diff
@@ -132,21 +134,132 @@ pre_configure() {
...
+	pushd webkit2gtk-41
+		configure_args=${configure_args//-DUSE_SOUP2=ON/}
+		configure_args=${configure_args//-DENABLE_WEBDRIVER=OFF/}
+		do_configure
+	popd
+	mkdir -p webkit2gtk-50
+	pushd webkit2gtk-50
+		configure_args+=" -DUSE_GTK4=ON -DENABLE_WEBDRIVER=OFF"
+		do_configure
+	popd
```
Comment (ahesford) [inline srcpkgs/webkit2gtk/template:148]:
```suggestion
	mkdir -p webkit2gtk-40 webkit2gtk-50
	(
		cd webkit2gtk-40 && \
		configure_args="${configure_args} -DUSE_SOUP2=ON -DENABLE_WEBDRIVER=OFF" \
		do_configure
	)
	
	(
		cd webkit2gtk-50 && \
		configure_args="${configure_args} -DUSE_GTK4=ON -DENABLE_WEBDRIVER=OFF" \
		do_configure
	)
```
Temporarily manipulating state like working directory or the value of a variable is undesirable and prone to error. Use subshells instead.

Also, because I modified the default args, it seems the default `do_build` run will produce `webkit2gtk41`, so I've swapped `webkit2gtk-41` for `webkit2gtk-40` here. You'll have to double-check my work.

File: srcpkgs/webkit2gtk/template [hunk 2]
```diff
@@ -132,21 +134,132 @@ pre_configure() {
...
+}
+
+pre_build() {
+	export NINJA_STATUS="[1/3][%f/%t] "
+}
```
Comment (ahesford) [inline srcpkgs/webkit2gtk/template:153]: Can this be moved to a top-level export to avoid another function? Or just leave it off; while the 2/3 and 3/3 prefixes are useful later, lack of a prefix can just imply 1/3.

File: srcpkgs/webkit2gtk/template [hunk 3]
```diff
@@ -132,21 +134,132 @@ pre_configure() {
...
+
+post_build() {
+	export NINJA_STATUS="[2/3][%f/%t] "
+	pushd webkit2gtk-41
+		do_build
+	popd
+	export NINJA_STATUS="[3/3][%f/%t] "
+	pushd webkit2gtk-50
+		do_build
+	popd
```
Comment (ahesford) [inline srcpkgs/webkit2gtk/template:163]:
```suggestion
	(cd webkit2gtk-40 && NINJA_STATUS="[2/3][%f/%t] " do_build)
	(cd webkit2gtk-50 && NINJA_STATUS="[3/3][%f/%t] " do_build)
```

File: srcpkgs/webkit2gtk/template [hunk 4]
```diff
@@ -132,21 +134,132 @@ pre_configure() {
...
+}
+
 post_install() {
+	pushd webkit2gtk-41
+		do_install
+	popd
+	pushd webkit2gtk-50
+		do_install
+	popd
+
```
Comment (ahesford) [inline srcpkgs/webkit2gtk/template:173]:
```suggestion
	(cd webkit2gtk-40 && do_install)
	(cd webkit2gtk-50 && do_install)
	
```

File: srcpkgs/webkit2gtk/template [hunk 5]
```diff
@@ -132,21 +134,132 @@ pre_configure() {
...
+	pkg_install() {
+		vmove usr/bin/WebKitWebDriver
+		vmove usr/libexec/webkit2gtk-4.1
+		vmove usr/lib/webkit2gtk-4.1
+		if [ "$build_option_gir" ]; then
+			vmove "usr/lib/girepository-1.0/*-4.1.typelib"
+		fi
+		vmove "usr/lib/*-4.1.so.*"
+	}
+}
```
Comment (ahesford) [inline srcpkgs/webkit2gtk/template:224]: This should just be in the standard `webkit2gtk` package. (I choose this one both because I modified the default configure args to produce, I believe, v4.1; and also because this seems to be the only subpackage that ships an executable in `usr/bin/WebKitWebDriver`. If you think one of the other two should be the "preferred" or "default" version, leave that one in the main `webkit2gtk` package.)

File: srcpkgs/webkit2gtk/template [hunk 6]
```diff
@@ -132,21 +134,132 @@ pre_configure() {
...
+		vmove "usr/lib/pkgconfig/*-4.1.pc"
+		if [ "$build_option_gir" ]; then
+			vmove "usr/share/gir-1.0/*-4.1.gir"
+			vmove "usr/share/gtk-doc/html/*-4.1"
+		fi
+		vmove "usr/lib/*-4.1.so"
+	}
+}
+
+webkit2gtk50_package() {
```
Comment (ahesford) [inline srcpkgs/webkit2gtk/template:240]: As above, this should probably be `libwebkit2gtk50` because its primary function is shipping a library (and its supporting content).

File: srcpkgs/webkit2gtk/template [hunk 7]
```diff
@@ -132,21 +134,132 @@ pre_configure() {
...
+		vmove usr/lib/webkit2gtk-4.0
+		vmove usr/libexec/webkit2gtk-4.0
+		if [ "$build_option_gir" ]; then
+			vmove "usr/lib/girepository-1.0/*-4.0.typelib"
+		fi
+		vmove "usr/lib/*-4.0.so.*"
+	}
+}
+
+webkit2gtk40-devel_package() {
```
Comment (ahesford) [inline srcpkgs/webkit2gtk/template:198]: This should probably be libwebkit2gtk40 because its primary function is shipping a library (and its supporting content).

File: srcpkgs/webkit2gtk/template (hunk 6 above)
Comment (oreo639) [reply srcpkgs/webkit2gtk/template:240]: All of their primary function is shipping a library, `WebDriver` is an language-agnostic api that allows you to control a browser, from say, a python script (i.e. for testing purposes):
https://developer.mozilla.org/en-US/docs/Web/WebDriver

So I don't see the need for the distinction. Feel free to disagree though.

File: srcpkgs/webkit2gtk/template (hunk 7 above)
Comment (oreo639) [reply srcpkgs/webkit2gtk/template:198]: See comment at webkit2gtk50

File: srcpkgs/webkit2gtk/template (hunk 1 above)
Comment (oreo639) [reply srcpkgs/webkit2gtk/template:148]: Thanks for the tip.

File: srcpkgs/webkit2gtk/template [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
 pkgname=webkit2gtk
-version=2.36.7
+version=2.38.0
 revision=1
 wrksrc="webkitgtk-${version}"
 build_style=cmake
 build_helper="gir"
 configure_args="-DPORT=GTK -DUSE_LD_GOLD=OFF
  -DCMAKE_LINKER=${XBPS_CROSS_TRIPLET}-gcc -DCMAKE_SKIP_RPATH=ON
  -DENABLE_JOURNALD_LOG=OFF -DUSE_WOFF2=ON -DUSE_SOUP2=ON
```
[the remark this answers is not in the document]
Comment (oreo639) [reply srcpkgs/webkit2gtk/template:10]: Alright.

**Comment by oreo639** [comment]
> It is not standard practice to split the `-devel` package by functionality.

We do this with python2 vs python3 packages. We also do this with libportal.
This would also allow us to split the packages (i.e. if we get tired of building one of them constantly and decide we can afford to update it less frequently)

File: srcpkgs/webkit2gtk/template (hunk 6 above)
Comment (ahesford) [reply srcpkgs/webkit2gtk/template:240]: The point is that we tend to name library-specific subpackages with `lib` prefixes. If there were a compelling reason to split out the (v4.1 in my modified structure) library from the main package, that should be `libwebkit2gtk41` as well. However, because there's no value in making `webkit2gtk` an empty package that just pulls in the (default) library, just keep that as `webkit2gtk`.

File: srcpkgs/webkit2gtk/template (hunk 5 above)
Comment (oreo639) [reply srcpkgs/webkit2gtk/template:224]: webkit2gtk40 is depended on by the webkit2gtk package in order to avoid having to rebuild all packages depending on webkit2gtk-4.0.
Currently the webkit2gtk package contains common files depended on by all subpackages (i.e. the locale files), although it might be better to have a webkit2gtk-common?
Imo, it's better to leave them separated, then we can decide which one is 

> also because this seems to be the only subpackage that ships an executable in `usr/bin/WebKitWebDriver`

That executable is meant to be used by say, C# or python applications for accessing the contents of webpages. (i.e. for testing purposes)
https://developer.mozilla.org/en-US/docs/Web/WebDriver

Similar to geckodriver (they both implement the same api).

**Comment by ahesford** [comment]
Save the devel split for when it's needed, don't complicate the package structure in anticipation of something that may never happen.

File: srcpkgs/webkit2gtk/template (hunk 5 above)
Comment (ahesford) [reply srcpkgs/webkit2gtk/template:224]: In light of that, the common stuff should definitely be moved into a common subpackage that each `lib*` package can depend on. The `webkit2gtk` package itself should could either contain whatever the "default" version is (if a sensible default can be defined).

Comment (oreo639) [reply srcpkgs/webkit2gtk/template:224]: Is it okay to be a metapackage pointing to the default package?

**Comment by oreo639** [comment]
One blocker is that running webkit2gtk in sandbox mode results in the web processes being unable to connect to dbus, this seems to be an issue on other non-systemd distros, not sure what the cause it.
We were able to ignore it, however the next release of epiphany will make it a hard error.

I don't run into that issue on flatpak though (i.e. the epiphany 43 flatpak).

**Comment by paper42** [comment]
> We were able to ignore it in the past, however the next release of epiphany will make it a hard error.

If this is an error only with epiphany 43, then we don't have to fix it in this PR. This PR is holding quite a bit of updates, so it would be nice to get this merged.

**Comment by oreo639** [comment]
> If this is an error only with epiphany 43, then we don't have to fix it in this PR.

TL;DR: webkit2gtk fails to connect to dbus since it disables networking in containers and the dbus default configuration uses tmpdir which defaults to abstract sockets (although there is a PR in dbus to change that). It doesn't affect distros using the systemd user init because that overrides the settings to use the filesystem instead of abstract sockets.
(This doesn't affect flatpak because flatpak creates its own socket path and uses it in the containerss which webkit2gtk is supposed to do but doesn't)

**Comment by ahesford** [comment]
Can't the socket issue be fixed with a configuration change in dbus?

**Comment by oreo639** [comment]
Yes, although if the configuration doesn't update it will still be broken.

**Comment by ahesford** [comment]
In that case, I think we resolve this problem by adding an INSTALL.msg to the relevant webkit2gtk package(s) noting that sandboxed applications may require this configuration change and leave it at that. (If you want to provide more details, *e.g.* examples, install a README.voidlinux and have the INSTALL.msg refer to it.)

We don't need to anticipate and resolve any upstream changes that will break without configuration updates; that's the job of the user. (It would be nice if we could mark some updates as "manual intervention required" to force some critical notices to receive user attention, but alas, XBPS provides us no suitable mechanism.)

**Comment by oreo639** [comment]
Alright, it should be fixed now.

File: srcpkgs/webkit2gtk/template [hunk 9]
```diff
@@ -142,11 +165,75 @@ webkit2gtk-devel_package() {
...
+		if [ "$build_option_gir" ]; then
+			vmove "usr/share/gir-1.0/*-4.0.gir"
+			vmove "usr/share/gtk-doc/html/*-4.0"
+		fi
+		vmove "usr/lib/*-4.0.so"
+	}
+}
+
+webkit2gtk-common_package() {
+	short_desc+="GTK port of the WebKit2 browser engine - common files"
```
Comment (paper42) [inline srcpkgs/webkit2gtk/template:179]:
```suggestion
	short_desc="GTK port of the WebKit2 browser engine - common files"
```

File: srcpkgs/webkit2gtk/template [hunk 10]
```diff
@@ -142,11 +165,75 @@ webkit2gtk-devel_package() {
...
+	pkg_install() {
+		vmove usr/share/locale/
+		vmove usr/share/licenses/
+	}
+}
+
+libwebkit2gtk41_package() {
+	depends="${depends}"
```
Comment (paper42) [inline srcpkgs/webkit2gtk/template:187]: I am not sure if this is safe, are you sure this will always use $depends from the main package? Either way, it's only webkit2gtk-common, so I think it might make sense to just write that.

File: srcpkgs/webkit2gtk/template [hunk 11]
```diff
@@ -132,7 +132,30 @@ pre_configure() {
...
+post_configure() {
+	mkdir -p webkit2gtk-40 webkit2gtk-50
+	(
+		cd webkit2gtk-40 && \
+		configure_args="${configure_args} -DUSE_SOUP2=ON -DENABLE_WEBDRIVER=OFF" \
+		do_configure
+	)
+
+	(
+		cd webkit2gtk-50 && \
```
Comment (paper42) [inline srcpkgs/webkit2gtk/template:144]:
```suggestion
		cd webkit2gtk-50
```
or is there a reason why we should use `&&` @ahesford?

File: srcpkgs/webkit2gtk/template [hunk 12]
```diff
@@ -20,23 +19,24 @@ configure_args="-DPORT=GTK -DUSE_LD_GOLD=OFF
...
  woff2-devel freetype-devel libopenjpeg2-devel libatomic-devel
  qt5-devel libmanette-devel libwpe-devel wpebackend-fdo-devel
  libgcrypt-devel libnuspell-devel libpsl-devel $(vopt_if x11 libXt-devel)
  $(vopt_if wayland 'MesaLib-devel libxkbcommon-devel wayland-devel wayland-protocols')"
+depends="webkit2gtk-common"
 short_desc="GTK+3 port of the WebKit2 browser engine"
 maintainer="Orphaned <orphan@voidlinux.org>"
 license="LGPL-2.1-or-later, BSD-2-Clause"
 homepage="https://webkitgtk.org/"
 distfiles="${homepage}/releases/webkitgtk-${version}.tar.xz"
```
Comment (paper42) [inline srcpkgs/webkit2gtk/template:38]:
```suggestion
distfiles="https://webkitgtk.org//releases/webkitgtk-${version}.tar.xz"
```

File: srcpkgs/webkit2gtk/template [hunk 13]
```diff
@@ -20,23 +19,24 @@ configure_args="-DPORT=GTK -DUSE_LD_GOLD=OFF
 # Don't remove which from hostmakedepends
 # Otherwise, they invoke /usr/bin/ccache /usr/lib/ccache/bin/$CC
 hostmakedepends="perl python pkg-config gperf flex ruby gettext glib-devel
```
Comment (paper42) [inline srcpkgs/webkit2gtk/template:21]: Can we use python3?

File: srcpkgs/webkit2gtk/template (hunk 11 above)
Comment (ahesford) [reply srcpkgs/webkit2gtk/template:144]: Not really, I'm just in the habit of using `&&` in these constructs to fail fast and avoid side effects.

Comment (paper42) [reply srcpkgs/webkit2gtk/template:144]: any error in the template should stop the build. If the subshell inherits that property, I think && could be removed.

File: srcpkgs/webkit2gtk/template (hunk 10 above)
Comment (oreo639) [reply srcpkgs/webkit2gtk/template:187]: It also needs bubblewrap and xdg-dbus-proxy

**Comment by paper42** [comment]
2.38.1 is released

**Comment by oreo639** [comment]
> 2.38.1 is released

Thanks. I tested it and it seems to work fine.
